providers/fips/selftest.c

```diff
@@ -85,14 +85,44 @@ BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
     }
     return TRUE;
 }
+#elif defined(__sun)
+
+void dep_init(void);
+void dep_cleanup(void);
+
+#pragma init(dep_init)
+void dep_init(void)
+{
+    FIPS_state = FIPS_STATE_SELFTEST;
+}
+
+#pragma fini(dep_cleanup)
+void dep_cleanup(void)
+{
+    CRYPTO_THREAD_lock_free(self_test_lock);
+}
+
+#elif defined(__hpux)
+
+#pragma init "dep_init"
+void dep_init(void)
+{
+    FIPS_state = FIPS_STATE_SELFTEST;
+}
+
+#pragma fini "dep_cleanup"
+void dep_cleanup(void)
+{
+    CRYPTO_THREAD_lock_free(self_test_lock);
+}
```

Comment:
Could we use init/cleanup names which are consistent with the GNUC version?

Also I wonder if we could somehow combine all these function implementations together somehow. They're all basically the same function implementations with variations around how we declare them or mark them as init/fini functions.

Reply:
Ok I have attempted to do the second thing..
you cant put #pragmas inside macros so it does become a bit messy...